hfile.c

```diff
@@ -527,6 +526,49 @@ typedef struct {
     unsigned is_socket:1;
 } hFILE_fd;
 
+/// hclose2 - closes the file using hclose when it is not stdin or stdout
+/** @param fp pointer to hFile
+ * returns 0 on success, -1 on failure
+*/
+int hclose2(hFILE *fp)
+{
+    hFILE_fd *hp = (hFILE_fd*)fp;
```

Comment:
This will absolutely not work. A public API function taking an `hFILE` may be called with any kind of hFILE (and will be called thus given e.g. `bgzip -dc s3://foo/bar.gz`), not just a local file that is represented as an `hFILE_fd`. This cast is invalid and in general there is no way to tell what flavour of `hFILE` you are being given, and most of them won't return anything meaningful for `hp->fd`.

Instead you will need to track or test in _bgzip.c_ whether the input/output file is stdin/out and avoid calling `bgzf_close`/`hclose` appropriately. Then you won't need to touch _bgzf.c_ or _hfile.[ch]_ at all.

Reply:
Yes, I think the solution for `stdout` would be to open it if needed outside the main processing loop,  and pass it into `operate()`.  Then you can keep writing to the same file handle.

For `stdin`, it's arguable that it shouldn't be allowed when specifying file names unless you're using `-c` as it would write to `-.gz` which is probably not what was intended.  (Although testing the existing `bgzip` it does do that...)

Reply:
It is admittedly annoying when you accidentally close your only connection to stdin/out. 😄 

However IMHO it is more important for `hclose()` (and `fd_close()`) to reliably actually close things.

In particular, there are some I/O errors that are not reported until you call `close(2)`. So to fully check for errors in `some-htslib-using-command args > output.bam` you need to check the return value of `hclose()` and it needs to reflect the return value of `close(2)` even for stdout.

Reply:
Just to avoid confusion, I assume you mean `close(2)` as in the system call `close` in section 2 of the man pages, rather than fd 2 (traditionally stderr).    Agreed though, we want to close stdout and if that produces an error report it to stderr via the usual channels.

I think this PR shouldn't need any htslib API changes.

When operating on a single file, it worked as before barring maybe some idioscyncracies like -.gz that ought to be tidied - gzip input stdin means output to stdout (but that's really an orthogonal issue).  When operating on multiple files it's really just a loop over the arguments with existing semantics assuming command line options are compatible.  Generally that means multiple files, not stdin/stdout.  The only slight fly in the ointment is attempt to specify "-" or "/dev/stdout" of one of a series of files.  I think that's such a weird corner case we probably don't want to let the tail wag the dog and just accept we may have closed stdin/stdout early and live with it. 

Curiously, this actually works though:

```
$ gzip - - > _.gz
foo
^D
bar
^D
$ zcat _.gz
foo
bar
```

That surprised me.  It's such an esoteric case though I really don't think it's worth trying to handle it.

Reply:
I am sure @jkbonfield is well aware that I'm old school enough that I'm using `manpage(section)` notation here despite it being annoyingly ambiguous for a function with one integer argument. :smile:

Seems to me (not having thought about it enormously much) one of the few use cases that would require keeping stdin or stdout open that's actually useful is `bgzip -dc foo.gz bar.gz baz.gz`.

Reply:
Obviously - it was more for @vasudeva8's benefit who may not be aware of the long-standarding unix notation.

Good point on the `bgzip -dc foo.gz bar.gz baz.gz` case.  Not closing stdout is far more important here than not closing stdin, which is a very esoteric terminal-only situation (I think).  This still doesn't need API changes though. We can just delay the htsclose of stdout until the last file has been written.

Reply:
updated based on comments